packages/headless/src/controllers/insight/user-actions/headless-user-actions.ts

```diff
@@ -0,0 +1,145 @@
+import {loadReducerError} from '../../../utils/errors';
+import {
+  buildController,
+  Controller,
+} from '../../controller/headless-controller';
+
+export interface UserActionsProps {
+  /**
+   * The options for the `UserActions` controller.
+   */
+  options: UserActionsOptions;
+}
+
+export interface UserActionsOptions {
+  /**
+   * The ticket creation date in the [ISO 8601](https://www.iso.org/iso-8601-date-and-time-format.html) format.
+   */
+  ticketCreationDate: string;
+  /**
+   * The number of sessions to include occuring before the ticket creation session.
+   * @defaultValue `0`
+   */
+  numberSessionsBefore?: number;
+  /**
+   * The number of sessions to include occuring after the ticket creation session.
+   * @defaultValue `0`
+   */
+  numberSessionsAfter?: number;
+  /**
+   * The names of custom events to exclude.
+   * @defaultValue `[]`
+   */
+  excludedCustomActions?: string[];
+}
+
+export interface UserActionsState {
+  /**
+   * The timeline of user actions.
+   */
+  timeline: UserActionTimeline;
+  /**
+   * The number of sessions included before the ticket creation date.
+   */
+  numberSessionsBefore: number;
+  /**
+   * The number of sessions included after the ticket creation date.
+   */
+  numberSessionsAfter: number;
+  /**
+   * The names of custom actions to exclude from the user actions.
+   */
+  excludedCustomActions: string[];
+  /**
+   * The ticket creation date in the [ISO 8601](https://www.iso.org/iso-8601-date-and-time-format.html) format.
+   */
+  ticketCreationDate?: string;
+  /**
+   * `true` if fetching the user actions is in progress and `false` otherwise.
+   */
+  loading: boolean;
+  /**
+   * The error response if fetching the user actions failed.
+   */
+  error?: InsightAPIErrorStatusResponse;
+}
+
+/**
+ * The UserActions controller is responsible for fetching user actions surrounding a given case creation event.
+ */
+export interface UserActions extends Controller {
+  /**
+   * Fetch the list of user actions surrounding the ticket creation.
+   */
+  fetchUserActions(): void;
+  /**
+   * Fetch the list of user actions including one more session preceding the ticket creation.
+   */
+  fetchWithPrecedingSession(): void;
+  /**
+   * Fetch the list of user actions including one more session following the ticket creation.
+   */
+  fetchWithFollowingSession(): void;
+  /**
+   * The state of the UserActions controller.
+   * */
+  state: UserActionsState;
+}
+
+export function buildUserActions(
+  engine: InsightEngine,
+  props: UserActionsProps
+): UserActions {
+  if (!loadUserActionsReducers(engine)) {
+    throw loadReducerError;
+  }
+
+  const {dispatch} = engine;
+  const getState = () => engine.state.insightUserAction;
+  const controller = buildController(engine);
+
+  dispatch(registerUserActions(props.options));
+
+  return {
+    ...controller,
+
+    get state() {
+      return getState();
+    },
+
+    fetchUserActions() {
+      dispatch(fetchUserActions());
```

Comment:
Will the list of user actions fetched by this action contain the `case creation` event? or we need to add this event manually based on the `ticketCreationDate` option?

Reply:
Good question. TBH I think the easiest is to omit the case creation event from the timeline given that a timestamp is passed to the API as the ticketCreationDate. Using a timestamp to fetch events from the database is pretty easy but trying to match the salesforce timestamp with a specific event in the database sounds like a headache as I'm not sure they would necessarily match. So the ticket creation event would be omitted and on the UI side we trust that the timestamp supplied is the correct one and use that to mark the event in the timeline.

But we wouldn't have the searchhub like we see in the mockups.

What do you think?

Reply:
@drosenfeld-coveo here's a discussion that's in line with [your question](https://github.com/coveo/ui-kit/pull/2637#discussion_r1062941455)

Reply:
This sounds good to me, we can handle the case creation event from the UI, we would still have the headache of trying to display the case creation event in the right order among all the user actions, but I think it is possible.

Reply:
we can also just completely omit any case creation event from the timeline. Thinking about it i don't see it being relevant to show creation of other cases and the main case creation event in the UI will just be created from the timestamp.

Reply:
Considering the API parameter is named `ticketCreationDate`, it would actually make sense to me to have the API return a case creation event matching the provided date. If the provided reference point was arbitrary (case creation date or something else), then inserting the case creation event would become problematic.

Does it makes sense to you?